File: Problems/1472. (M) Design Browser History/Java/solution.java
class BrowserHistory {
  // doubly linked list

  private class HistoryNode {
    String url;
    HistoryNode prev;
    HistoryNode next;

    public HistoryNode(String url) {
      this.url = url;
      this.prev = null;
      this.next = null;
    }

    public HistoryNode(String url, HistoryNode p) {
      this.url = url;
      this.prev = p;
      this.next = null;
    }
  }

  HistoryNode cur;

  public BrowserHistory(String homepage) {
    this.cur = new HistoryNode(homepage);
  }
  
  public void visit(String url) {
    // unlink next node and attach new with url url
    if (this.cur.next != null)
      this.cur.next.prev = null;
    this.cur.next = new HistoryNode(url, this.cur);
    this.cur.next.prev = this.cur;
    this.cur = this.cur.next;
  }
  
  public String back(int steps) {
    // move back steps nodes or until first node is reached
    while (steps > 0 && this.cur.prev != null) {
      this.cur = this.cur.prev;
      steps--;
    }
    return this.cur.url;
  }
  
  public String forward(int steps) {
    // move forward steps nodes or until last node is reached
    while (steps > 0 && this.cur.next != null) {
      this.cur = this.cur.next;
      steps--;
    }
    return this.cur.url;
  }
}

File: Problems/1472. (M) Design Browser History/Python3/solution.py
class BrowserHistory:
  # doubly linked list

  class _HistoryNode:
    def __init__(self, url, p = None, n = None):
      self.url = url
      self.prev = p
      self.next = n

  def __init__(self, homepage: str):
    self.cur = self._HistoryNode(homepage)

  def visit(self, url: str) -> None:
    # unlink next node and attach new with url url
    if self.cur.next:
      self.cur.next.prev = None
    self.cur.next = self._HistoryNode(url, self.cur)
    self.cur.next.prev = self.cur
    self.cur = self.cur.next

  def back(self, steps: int) -> str:
    # move back steps nodes or until first node is reached
    while steps and self.cur.prev:
      self.cur = self.cur.prev
      steps -= 1
    return self.cur.url

  def forward(self, steps: int) -> str:
    # move forward steps nodes or until last node is reached
    while steps and self.cur.next:
      self.cur = self.cur.next
      steps -= 1
    return self.cur.url

